web/pandas/pdeps/0010-required-pyarrow-dependency.md

```diff
@@ -0,0 +1,138 @@
+- Author: [Matthew Roeschke](https://github.com/mroeschke)
+          [Patrick Hoefler](https://github.com/phofl)
+- Revision: 1
+
+## Abstract
+
+This PDEP proposes that:
+
+- PyArrow becomes a runtime dependency starting with pandas 2.1
+- The minimum version of PyArrow supported starting with pandas 2.1 is version 7 of PyArrow.
+- When the minimum version of PyArrow is bumped, PyArrow will be bumped to the highest version that has
+  been released for at least 2 years.
+
+## Background
+
+PyArrow is an optional dependency of pandas that provides a wide range of supplemental features to pandas:
+
+- Since pandas 0.21.0, PyArrow provided I/O reading functionality for Parquet
+- Since pandas 1.2.0, pandas integrated PyArrow into the `ExtensionArray` interface to provide an
+  optional string data type backed by PyArrow
+- Since pandas 1.4.0, PyArrow provided I/0 reading functionality for CSV
+- Since pandas 1.5.0, pandas provided an `ArrowExtensionArray` and `ArrowDtype` to support all PyArrow
+  data types within the `ExtensionArray` interface
+- Since pandas 2.0.0, all I/O readers have the option to return PyArrow-backed data types, and many methods
+  now utilize PyArrow compute functions to
+accelerate PyArrow-backed data in pandas, notibly string and datetime types.
+
+As of pandas 2.0, one can feasibly utilize PyArrow as an alternative data representation to NumPy with advantages such as:
+
+1. Consistent ``NA`` support for all data types
+2. Broader support of data types such as ``decimal``, ``date`` and nested types
+
+## Motivation
+
+While all the functionality described in the previous paragraph is currently optional, PyArrow has significant
+integration into many areas of pandas. With our roadmap noting that pandas strives for better Apache Arrow
+interoperability [^1] and many projects [^2], within or beyond the Python ecosystem, adopting or interacting with
+the Arrow format, making PyArrow a required dependency provides an additional signal of confidence in the Arrow
+ecosystem to pandas users.
+
+Additionally, requiring PyArrow would simplify the related development within pandas and potentially improve NumPy
+functionality that would be better suited by PyArrow including:
+
+- Avoiding runtime checking if PyArrow is available to perform PyArrow object inference during constructor or indexing operations
+  - Currently, there are 17 `import_optional_dependency("pyarrow")` checks throughout the pandas code base
+
+- Removing unnecessary functionality:
+  - fastparquet engine in ``read_parquet``
+  - potentially simplifying the ``read_csv`` logic (needs more investigation)
+
+- Avoiding NumPy object data types more by default for analogous types that have native PyArrow support such as:
+  - decimal
+  - binary
+  - nested types (like lists, dicts, ...)
+  - strings
+
+Out of this group, strings offer the most advantages for users. They use significantly less memory and are faster:
```

Comment:
Haven't kept up with this, but how are the plans to add the new numpy string dtype (xref #47884 ) going to affect the rationale here?

I would assume performance of the numpy string dtype would be on par with the pyarrow one.

Reply:
The linked comment said that numpy strings are available "within a year or so". 
This does not seem to be dissimilar to a pandas 3.0 release date now proposed here?

Reply:
> this is still years' away

I hope it doesn't take that long!

> (Nathan - we discussed timelines before, but I didn't write them down so have forgotten them, apologies)

The earliest pandas could officially support the dtype I'm working on is after the release of Numpy 2.0 - currently scheduled for January 2024. This assumes the new dtype API is available for downstream use in Numpy 2.0 without needing to set an environment variable. I'm hoping to start shipping experimental support in pandas behind the environment variable after Numpy 1.25 is released this summer, as that version of Numpy will hopefully have a version of the new dtype API that is usable for pandas' needs. The version in Numpy 1.24 is broken and is missing a lot of features we've added since that release, unfortunately.

Reply:
> I can't remember the perf comparison - @ngoldbaum do you want to comment here?

Creating e.g. a `Series` or `Dataframe` with `dtype=StringDtype()` is now substantially fasterin my pandas fork than with object strings and pyarrow strings, since there's no overhead from the ExtensionArray interface. I've also implemented an `ExtensionArray` (mostly so I can use the extension tests in pandas to ferret out places in numpy that need to be fixed), and that has performance comparable to object strings and pyarrow strings.

String method operations (e.g. `series.str.capitalize()`) have performance comparable to object strings but have worse performance than pyarrow strings where a fast implementation in pyarrow is available. In principle numpy could match performance by implementing string operations as ufuncs, but for now I'm punting on that since it will require bikeshedding a public API and figuring out whether numpy should depend ICU, a lightweight unicode library, or should implement core unicode algorithms from scratch.

Memory usage is improved compared with object strings and pyarrow strings, and substantially improved compared with numpy fixed-width unicode strings, especially in worst-case situations like many short strings and one very long strings or mostly ASCII data. That's measured using `asv`'s peakmem benchmarks, which use the maximum resident size for the process as reported by the kernel, so it includes the binary size. Most of the memory usage in the pandas benchmarks is the binaries imported by pandas and the differences in memory usage in the benchmarks are much smaller than the total memory used by the process. Unfortunately more fine-grained benchmarking is very difficult as only some allocations in all these libraries participate in python's allocation tracking machinery.

Reply:
Thanks for explaining

Algorithm-wise, how mature is it? Do you think it could be a drop-in alternative to `pyarrow[string]`, or is there anything not implemented / that would work differently / which hasn't been extensively tested?

Reply:
I have a fork of pandas where I’ve added a `string[numpy]` ExtensionArray based on my dtype. It passes the full ExtensionArray test suite and almost passes all of the additional string array tests, the remaining failures are from a missing cast to/from datetimes (I’m working on support for that) and converting to/from PyArrow arrays, which I just haven’t looked at yet.

My hope is that it will be a drop-in replacement for `string[python]`, any ways in which it behaved differently I’d consider a bug. I’m not aware of any fundamental issues that would block making it a transparent drop-in replacement.

Right now I’m working on a NEP to propose adding the dtype to NumPy for NumPy 2.0 at the end of this year.

Reply:
IMO the efficient algorithms that PyArrow brings are even more important for us than simply the ``string[python]`` behavior. How does the ExtensionArray on your fork compare to the ``string[pyarrow]`` implementation? Performance and memory usage would be interesting, although I am assuming that there shouldn't be a huge memory difference?

Reply:
> How does the ExtensionArray on your fork compare to the string[pyarrow] implementation?

The memory usage should be comparable with pyarrow strings. Both are storing UTF8 bytestreams internally. I don't know offhand if arrow uses the small string optimization (storing the string content in the space normally reserved for a pointer to the string). It's difficult to compare memory usage exactly since the operating system facilities for this only allow you to measure the peak memory usage of a process and not all allocations necessarily use Python's allocation tracking machinery. I'm hoping to do a more careful memory usage benchmark as part of the NEP I'm writing.

The main difference in the storage is that right now I'm using individual heap allocations for each string array entry. Arrow just does a single allocation for all the array entries and has a secondary array of offsets to find the data for each string array element. I've thought a bit about following that approach, but it would mean we would have to either disallow mutating string arrays or there or have pathological behavior where enlarging a single array element could cause the entire array to get reallocated. It would also be nice to be able to use the short string optimization where a bitflag indicates whether an array element contains a pointer to heap data or contains a short string that is only a few bytes long, eliminating the need for a heap allocation for short strings.

For performance, do you mean for string manipulation operations like case folding or padding? In principle NumPy could add string ufuncs that would allow for fast implementations, but right now NumPy doesn't have a namespace for that. Currently, all the comparison operators are implemented as ufuncs, but no other string functionality is. There are string manipulation functions in the `np.char` namespace, but they just do a for loop over the array elements and call string functions on the scalars. 

I don't want to promise that string ufuncs definitely will happen in the future, but there's no real technical blockers, just social ones. NumPy doesn't currently have any ufuncs that only make sense with string data, so some thought needs to go into where in the namespace they should go. It will also require a decent amount of implementation work to add the functions, although mostly just tedious C coding.

Overall the goal is to facilitate a straightforward transition from workflows that used object string arrays while enabling possible performance improvements in the future that are currently impossible with object string arrays.

Reply:
Generally, pandas is moving away from mutability in some sense (the CoW adoption), so that isn't very high on my priority list.

While a storage efficient string dtype is nice, this is kind of pointless if the operations aren't fast from a pandas PoV. One of the biggest advantages of arrow is that we can reduce memory but also that most operations are significantly faster, depending on what you are doing it's can be an order of magnitude. 

I am referring mostly to stuff like the ``str`` accessor but also things like factorization etc. 

So even if NumPy strings are ready in around a year (or some other time period), that's not helpful for us as long as NumPy does not ship fast algorithms on top of it. 

Sorry if this sounds harsh, that wasn't my intention. But having the string dtype without algorithms gets us only half the way compared to what PyArrow does, so this isn't a compelling argument to avoid making Arrow strings the default.

Reply:
As a minimum, a fast regex engine could potentially help as some of the str accessor functions were (maybe still are) implemented using regex for string[pyarrow] where the functions did not exist in PyArrow (or the minimum version supported at the time).

Reply:
Are you suggesting to implement this in pandas? That's something I personally don't have any interest in doing and would also be at least 0- on adding for the time being. Having this stuff in Arrow is nice since it reduces maintenance burden and also having better test coverage since more libraries will depend on it

Reply:
I suspect a regex engine would be implemented in Numpy and then any str accessor functions not implemented in NumPy could be implemented using either regex or object fallback in pandas (just like we did for PyArrow initially).